Check username

kaapstorm opened the pull request:
## Technical Summary

Context: https://github.com/dimagi/commcare-hq/pull/31482#pullrequestreview-944316229

For authentication types that require a username, raise `ValueError` if it isn't set.


## Safety Assurance

### Safety story

Behavior change: For any remote API connections that use an authentication type that requires a username, if the username wasn't given it would be failing authentication, because the username would be sent as "None". In other words, this change can't break anything that isn't already broken.

### Automated test coverage

Code change covered by tests.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
Raise ValueError on bad username

kaapstorm committed:
Fix test

kaapstorm commented on corehq/motech/models.py:
```diff
@@ -153,8 +153,29 @@ def get_requests(
+            )
+
+        # Auth types that require a username:
+        if not isinstance(self.username, str):
```
I considered checking for `None`, or just whether `self.username` is Truthy, but chose explicitly checking it's a `str` because that's what's breaking.

kaapstorm committed:
Add form validation

kaapstorm commented:
Commit [f7e9625](https://github.com/dimagi/commcare-hq/pull/31487/commits/f7e962556bf4d1ec2b5321c835f253c3c3e8851d) adds form validation. It's not as good as a DB constraint, but it's an easy lift that catches problems at the start.

![image](https://user-images.githubusercontent.com/708421/163816183-77e4c839-6f1f-4ca1-9884-3aa39eeb4379.png)

millerdev commented on corehq/motech/models.py:
```diff
@@ -153,8 +153,29 @@ def get_requests(
+
+        # Auth types that require a username:
+        if not isinstance(self.username, str):
+            raise ValueError(f'Auth type {self.auth_type!r} requires a username')
```
Is it possible that `ConnectionSettings` exist in production databases that will now raise an error here? It seems like it would be a problem if that happened. What do you think of changing this until the `CHECK` constraint solution can be implemented (with accompanying migration to fix bad rows, assuming that is possible)?
```suggestion
            return AuthManager()
```

kaapstorm replied:
Although those `ConnectionsSettings` would never have worked, you're right, it is possible that they exist (and almost guaranteed to exist in a QA domain on Staging). I'll change this.

millerdev reviewed (commented):
Can this PR also remove the `filter_warnings()` in [corehq/apps/export/tests/test_incremental.py](https://github.com/dimagi/commcare-hq/commit/e52ea875ab36c59b8f5a70d7219846da2e66605d#diff-55efddca710caac4d8c54c51835dddfd9b03323e0f2537baad78b9a2bc084df8)?

kaapstorm committed:
return AuthManager()

Authentication will fail without a username. Form validation and a DB constraint should prevent this from happening.

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

millerdev committed:
Merge branch 'master' into nh/check_username

millerdev committed:
Remove obsolete warning filter

kaapstorm commented on corehq/apps/export/tests/test_incremental.py:
```diff
@@ -179,19 +179,10 @@ def test_sending_fail(self):
-                module="requests.auth",
-            ):
-                _send_incremental_export(self.incremental_export, checkpoint)
+            _send_incremental_export(self.incremental_export, checkpoint)
```
Thank you
